src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -177,124 +196,169 @@ export interface SavedObjectsSetupDeps {
   elasticsearch: InternalElasticsearchServiceSetup;
 }
 
+interface WrappedClientFactoryWrapper {
+  priority: number;
+  id: string;
+  factory: SavedObjectsClientWrapperFactory<KibanaRequest>;
+}
+
 /** @internal */
 // eslint-disable-next-line @typescript-eslint/no-empty-interface
 export interface SavedObjectsStartDeps {}
 
 export class SavedObjectsService
   implements CoreService<InternalSavedObjectsServiceSetup, InternalSavedObjectsServiceStart> {
-  private migrator: KibanaMigrator | undefined;
   private logger: Logger;
-  private clientProvider: ISavedObjectsClientProvider<KibanaRequest> | undefined;
+
+  private setupDeps?: SavedObjectsSetupDeps;
+  private clientFactoryProvider?: SavedObjectsClientFactoryProvider<KibanaRequest>;
+  private clientFactoryWrappers: WrappedClientFactoryWrapper[] = [];
+
+  private mappings: SavedObjectsMapping[] = [];
+  private migrations: MigrationDefinition = {};
+  private schemas: SavedObjectsSchemaDefinition = {};
+  private validations: PropertyValidators = {};
 
   constructor(private readonly coreContext: CoreContext) {
     this.logger = coreContext.logger.get('savedobjects-service');
   }
 
-  public async setup(
-    setupDeps: SavedObjectsSetupDeps,
-    migrationsRetryDelay?: number
-  ): Promise<InternalSavedObjectsServiceSetup> {
+  public async setup(setupDeps: SavedObjectsSetupDeps): Promise<InternalSavedObjectsServiceSetup> {
     this.logger.debug('Setting up SavedObjects service');
 
+    this.setupDeps = setupDeps;
+
     const {
       savedObjectSchemas: savedObjectsSchemasDefinition,
       savedObjectMappings,
       savedObjectMigrations,
       savedObjectValidations,
     } = setupDeps.legacyPlugins.uiExports;
+    this.mappings = savedObjectMappings;
+    this.migrations = savedObjectMigrations;
+    this.schemas = savedObjectsSchemasDefinition;
+    this.validations = savedObjectValidations;
 
-    const savedObjectSchemas = new SavedObjectsSchema(savedObjectsSchemasDefinition);
+    return {
+      setClientFactoryProvider: provider => {
+        if (this.clientFactoryProvider) {
+          throw new Error(
+            'custom client factory is already set, unable to replace the current one'
```

Comment:
Optional: `unable to replace the current one` raises the `why` question. should we clarify that `custom client factory can be set only once.` Or to shorten it to `custom client factory was already set`?